utils/v4l2-tracer/v4l2-tracer.1.in
.TH "V4L2-TRACER" "1" "November 2022" "v4l-utils @MAN_PACKAGE_VERSION@" "User Commands"
.SH NAME
v4l2-tracer - An application to trace and replay stateless video decoding.
.SH SYNOPSIS
\fBv4l2-tracer \fR[options] \fBtrace\fR <\fItracee\fR>
.RS
.RE
\fBv4l2-tracer \fR[options] \fBretrace\fR  <\fItrace_file\fR>\fB.json\fR
.RS
.RE
\fBv4l2-tracer clean\fR  <\fIfile\fR>\fB.json\fR
.RS
.RE

.SH DESCRIPTION
The v4l2-tracer utility traces, records and replays userspace applications
that implement the v4l2 memory-to-memory stateless video decoder interface.
.SS Trace
Trace system calls and video frame data passed by userspace application <\fItracee\fR> to kernel driver.
All stateless codec controls in user-space API can be traced. Outputs a JSON-formatted trace file.
.SS Retrace
Read the JSON-formatted <\fItrace_file\fR>\fB.json\fR. Replay the same system calls and pass the same video frame data to kernel driver.
Outputs a JSON-formatted retrace file.

.SS Clean
Remove lines with irrelevant differences (e.g. file descriptors and memory addresses) from JSON files.
Outputs a clean copy, not necessarily still in JSON-format.

.SH OPTIONS
.SS Common Options
.TP
\fB\-c\fR, \fB\-\-compact\fR
Write minimal whitespace in JSON file.
.TP
\fB\-g\fR, \fB\-\-debug\fR
Turn on verbose reporting plus additional debug info.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display this message.
.TP
\fB\-r\fR, \fB\-\-raw\fR
Write decoded video frame data to JSON file.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Turn on verbose reporting.
.TP
\fB\-y\fR, \fB\-\-yuv\fR
Write decoded video frame data to yuv file.

.SS Retrace Options
.TP
\fB\-d\fR, \fB\-\-device\fR <\fIdev\fR>
Use a different video device than specified in the trace file.
.RS
<\fIdev\fR> must be a digit corresponding to an existing /dev/video<\fIdev\fR>
.RE
.TP
\fB\-m\fR, \fB\-\-media\fR <\fIdev\fR>
Use a different media device than specified in the trace file.
.RS
<\fIdev\fR> must be a digit corresponding to an existing /dev/media<\fIdev\fR>
.RE

.SH EXIT STATUS
On success, it returns 0. Otherwise, it will return 1 or an error code.

.SH EXAMPLE
.TP
Trace an application decoding VP8 video:
.EX
\fIv4l2-tracer trace gst-launch-1.0 -- filesrc location=test-25fps.vp8 ! parsebin ! v4l2slvp8dec ! videocodectestsink\fR
.EE
.TP
A trace file is generated:
.EE
\fI71827_trace.json\fR
.TP
Retrace the trace file:
.EX
\fIv4l2-tracer retrace 71827_trace.json\fR
.EE
.TP
Specify device nodes if retracing on a different driver:
.EX
\fIv4l2-tracer -d0 -m0 retrace 71827_trace.json\fR
.EE
.TP
A retrace file is generated:
.EE
\fI71827_trace_retrace.json\fR
.EX
.TP
Remove file descriptors and addresses (optional):
.EX
\fIv4l2-tracer clean 71827_trace.json\fR
.EE
.EX
\fIv4l2-tracer clean 71827_trace_retrace.json\fR
.EE
.TP
Clean files are generated for comparison:
.EX
\fIclean_71827_trace.json\fR
.EE
.EE
\fIclean_71827_trace_retrace.json\fR
.EX

.SH BUGS
Bug reports or questions about this utility should be sent to the
linux-media@vger.kernel.org mailinglist.


ChangeLog
v4l-utils-1.26.1
----------------

Hans Verkuil (2):
      cec-ctl: fix random cec stress test: check if msg status is OK
      libv4l2rds: fix TMC location parsing

Peter Seiderer (3):
      media-info: add missing sys/stat.h include (for dev_t)
      meson: fix has_function fork detection (needs suitable include)
      meson: libv4lconvert helpers need fork support

Rudi Heitbaum (1):
      keytable: meson not passing rc_keymaps system directory

Sean Young (2):
      ir-ctl: remove line length limits
      keytable: remove line length limits


v4l-utils-1.26.0
----------------

Ariel D'Alessandro (3):
      Add support for meson building
      Copy Doxygen configuration file to doc/
      meson: Add support for doxygen documentation

Benjamin Gaignard (1):
      v4l2-compliance: Test queue maximum buffers allocation

Daniel Lundberg Pedersen (1):
      v4l2-ctl: Add --get/--set-ctrl support for INTEGER and INTEGER64 arrays

Deborah Brouwer (18):
      v4l2-tracer: print only if debugging option is set
      v4l2-tracer: add macros to print debug info
      v4l2-tracer: stop stringifying v4l2_input "tuner"
      v4l2-tracer: add G/S TUNER ioctls
      v4l2-tracer: add VIDIOC_ENUM_FRAMESIZES
      v4l2-tracer: add VIDIOC_ENUM_FRAMEINTERVALS
      v4l2-tracer: stringify target in v4l2_selection
      v4l2-tracer: add event tracing/retracing
      v4l2-compliance: call select before dequeuing event
      v4l2-tracer: fix autogen script for AV1
      v4l2-info/v4l2-tracer: add macro to mark the trace
      v4l2-tracer: replace buftype2s with val2s
      v4l2-tracer: remove buffers by type and index
      v4l2-tracer: remove compress_frame_count
      v4l2-tracer: get decoded bytesused from DQBUF
      v4l2-tracer: create an option to trace userspace args
      v4l2-tracer: stop retracing failed ioctls
      v4l2-tracer: auto generate flags for DECODER_CMD

Dikshita Agarwal (1):
      v4l2-compliance: Add handling for QC08C and QC10C format

Gregor Jasny (4):
      buildsystem: Start v4l-utils 1.25.0 development cycle
      build: conditionally retrieve Git repo information
      build: do not fall back to libbpf header lookup
      manpages: fix groff warnings

Hans Verkuil (53):
      v4l2-tracer: use __s64 instead of long
      v4l2-ctl: improve --stream-sleep functionality
      v4l2-ctl: reset fps counter when using --stream-sleep with streamoff/on
      v4l2-ctl: support random sleep times
      cec-ctl: store state changes in pin file
      configure.ac: fix gconv directory search
      v4l2-ctl: improve --stream-sleep
      sync-with-kernel.sh: regenerate utils/keytable/rc_keymaps/meson.build
      .gitignore: add back '*.pc' pattern
      v4l-utils: sync with upstream media staging kernel.
      gen_dvb_structs.pl: drop copy of dvb-frontend.h
      cec-ctl.1.in: clarify that you configure *your* device
      utils/libcecutil: drop spurious dep_libv4lconvert
      cec-ctl: add --test-random-power-states stress test
      cec/v4l2-compliance, cec-follower: use __FILE_NAME__ if available
      v4l2-tracer: add quotes to -DLIBTRACER_PATH="@0@"
      v4l2-ctl: show field value in print_concise_buffer()
      cec-ctl: rename power-cycle to standby-wakeup-cycle
      v4l-utils: sync with latest media staging tree
      cec-ctl: always log timestamp in show_msg if not-OK
      Update old email address
      cec-compliance: get_power_status(): fix valid RX test
      cec-compliance: get_power_status(): check cec_msg_status_is_abort
      v4l-utils: libdvbv5: fix doxygen warnings
      libcecutil: fix performance inefficiencies
      lib/libdvbv5/dvb-file.c: check for valid entry pointer
      media-ctl: 1 -> 1ULL
      v4l2-ctl: add missing > in aspect option usage message
      v4l2-ctl: fix handling the premul-alpha sub-option
      v4l2-ctl: improve --silent support
      cec-ctl: add hpd-may-be-low suboption for the CEC tests
      cec-ctl.1.in: for hpd-may-be-low option: <0,1> -> <0/1>
      test-media: cec-pwr enables cec variable as well
      v4l2-ctl/compliance: add include cstdint
      test-media: log utils versions
      cec-follower: add --exclusive option
      common/media-info: add MEDIA_ENT_F_PROC_VIDEO_ISP
      common/media-info: fail for unknown entity function values
      sync-with-kernel.sh: don't filter V4L2_PIX_FMT_HEVC_SLICE
      v4l-utils: sync with latest kernel headers
      v4l2-ctl: add support for AV1 stateless control types
      v4l2-compliance: support V4L2_PIX_FMT_AV1_FRAME
      v4l2-compliance: pass media_fd when walking topology
      test-media: 'mc' argument didn't select vidtv
      test-media: use -m instead of -M for vicodec and vidtv
      README.md: drop qt5-default, add libsdl2-dev libbpf-dev llvm clang
      v4l2-ctl: fix alignment meta formats usage message
      v4l2-ctl.1.in: add missing --help-meta
      v4l2-compliance: improve error message
      v4l-utils: sync with latest kernel headers
      v4l-utils: sync with latest kernel headers
      v4l2-info.cpp: fix _SUPPORTS_MAX_NUM_BUFFERS description
      v4l-utils: sync with latest kernel headers

Jorge Maidana (1):
      qv4l2: fix SDR format error message

Laurent Pinchart (3):
      Drop autoconf/automake support
      meson: Fix libbpf detection
      meson: Fix install location of doxygen's man pages

Mauro Carvalho Chehab (4):
      libdvbv5: dvb-scan: fix DVB-S2 checks for DVB-S tuning
      meson.build: backport it to version 0.56
      v4l-utils.spec.in: update it to work with meson
      meson: make it backward-compatible with version 0.54

Michael Tretter (1):
      v4l2-compliance: Add gadget. as bus_info prefix

Nicolas Dufresne (1):
      v4l2-tracer: Fix libv4l2tracer.so loader

Paul Kocialkowski (1):
      v4l2-compliance: codecs: Add stateless (TRY_)DECODER_CMD tests

Peter Seiderer (19):
      meson: re-enable qv4l2 without Qt OpenGL support
      qv4l2: remove leagcy alsa_stream.h include (ENABLE_ALSA protected)
      qv4l2: fix qmake project file

utils/v4l2-tracer/v4l2-tracer.cpp
/* SPDX-License-Identifier: GPL-2.0-only */
/*
 * Copyright 2022 Collabora Ltd.
 */

#include "retrace.h"
#include <climits>
#include <sys/stat.h>
#include <sys/wait.h>
#include <time.h>

int tracer(int argc, char *argv[], bool retrace = false);

pid_t tracee_pid = 0;

void v4l2_tracer_sig_handler(int signum)
{
	line_info("\n\tReceived signum: %d", signum);

	/* If there is a tracee, wait for it to handle the signal first before exiting. */
	if (tracee_pid) {
		kill(tracee_pid, signum);
		wait(nullptr);
	}
}

enum Options {
	V4l2TracerOptCompactPrint = 'c',
	V4l2TracerOptSetVideoDevice = 'd',
	V4l2TracerOptDebug = 'g',
	V4l2TracerOptHelp = 'h',
	V4l2TracerOptSetMediaDevice = 'm',
	V4l2TracerOptWriteDecodedToJson = 'r',
	V4l2TracerOptTraceUserspaceArg = 'u',
	V4l2TracerOptVerbose = 'v',
	V4l2TracerOptWriteDecodedToYUVFile = 'y',
};

const static struct option long_options[] = {
	{ "compact", no_argument, nullptr, V4l2TracerOptCompactPrint },
	{ "video_device", required_argument, nullptr, V4l2TracerOptSetVideoDevice },
	{ "debug", no_argument, nullptr, V4l2TracerOptDebug },
	{ "help", no_argument, nullptr, V4l2TracerOptHelp },
	{ "media_device", required_argument, nullptr, V4l2TracerOptSetMediaDevice },
	{ "raw", no_argument, nullptr, V4l2TracerOptWriteDecodedToJson },
	{ "userspace", no_argument, nullptr, V4l2TracerOptTraceUserspaceArg},
	{ "verbose", no_argument, nullptr, V4l2TracerOptVerbose },
	{ "yuv", no_argument, nullptr, V4l2TracerOptWriteDecodedToYUVFile },
	{ nullptr, 0, nullptr, 0 }
};

const char short_options[] = {
	V4l2TracerOptCompactPrint,
	V4l2TracerOptSetVideoDevice, ':',
	V4l2TracerOptDebug,
	V4l2TracerOptHelp,
	V4l2TracerOptSetMediaDevice, ':',
	V4l2TracerOptWriteDecodedToJson,
	V4l2TracerOptTraceUserspaceArg,
	V4l2TracerOptVerbose,
	V4l2TracerOptWriteDecodedToYUVFile
};

int get_options(int argc, char *argv[])
{
	int option = 0;

	do {
		/* If there are no commands after the valid options, return err. */
		if (optind == argc) {
			print_usage();
			return -1;
		}

		/* Avoid reading the tracee's options. */
		if ((strcmp(argv[optind], "trace") == 0) || (strcmp(argv[optind], "retrace") == 0))
			return 0;

		option = getopt_long(argc, argv, short_options, long_options, NULL);
		switch (option) {
		case V4l2TracerOptCompactPrint: {
			setenv("V4L2_TRACER_OPTION_COMPACT_PRINT", "true", 0);
			break;
		}
		case V4l2TracerOptSetVideoDevice: {
			std::string device_num = optarg;
			try {
				std::stoi(device_num, nullptr, 0);
			} catch (std::exception& e) {
				line_info("\n\tCan't convert <dev> \'%s\' to integer.", device_num.c_str());
				return -1;
			}
			if (device_num[0] >= '0' && device_num[0] <= '9' && device_num.length() <= 3) {
				std::string path_video = "/dev/video";
				path_video += optarg;
				setenv("V4L2_TRACER_OPTION_SET_VIDEO_DEVICE", path_video.c_str(), 0);
			} else {
				line_info("\n\tCan't use device number\'%s\'", device_num.c_str());
				return -1;
			}
			break;
		}
		case V4l2TracerOptDebug:
			setenv("V4L2_TRACER_OPTION_VERBOSE", "true", 0);
			setenv("V4L2_TRACER_OPTION_DEBUG", "true", 0);
			break;
		case V4l2TracerOptHelp:
			print_usage();
			return -1;
		case V4l2TracerOptSetMediaDevice: {
			std::string device_num = optarg;
			try {
				std::stoi(device_num, nullptr, 0);
			} catch (std::exception& e) {
				line_info("\n\tCan't convert <dev> \'%s\' to integer.", device_num.c_str());
				return -1;
			}
			if (device_num[0] >= '0' && device_num[0] <= '9' && device_num.length() <= 3) {
				std::string path_media = "/dev/media";
				path_media += optarg;
				setenv("V4L2_TRACER_OPTION_SET_MEDIA_DEVICE", path_media.c_str(), 0);
			} else {
				line_info("\n\tCan't use device number\'%s\'", device_num.c_str());
				return -1;
			}
			break;
		}
		case V4l2TracerOptWriteDecodedToJson:
			setenv("V4L2_TRACER_OPTION_WRITE_DECODED_TO_JSON_FILE", "true", 0);
			break;
		case V4l2TracerOptTraceUserspaceArg:
			setenv("V4L2_TRACER_OPTION_TRACE_USERSPACE_ARG", "true", 0);
			break;
		case V4l2TracerOptVerbose:
			setenv("V4L2_TRACER_OPTION_VERBOSE", "true", 0);
			break;
		case V4l2TracerOptWriteDecodedToYUVFile:
			setenv("V4L2_TRACER_OPTION_WRITE_DECODED_TO_YUV_FILE", "true", 0);
			break;
		default:
			break;
		}

		/* invalid option */
		if (optopt > 0) {
			print_usage();
			return -1;
		}

	} while (option != -1);

	return 0;
}

int clean(std::string trace_filename)
{
	FILE *trace_file = fopen(trace_filename.c_str(), "r");
	if (trace_file == nullptr) {
		line_info("\n\tCan't open \'%s\'", trace_filename.c_str());
		return 1;
	}

	fprintf(stderr, "Cleaning: %s\n", trace_filename.c_str());

	std::string clean_filename = "clean_" + trace_filename;
	FILE *clean_file = fopen(clean_filename.c_str(), "w");
	if (clean_file == nullptr) {
		line_info("\n\tCan't open \'%s\'", clean_filename.c_str());
		return 1;
	}

	std::string line;
	char buf[SHRT_MAX];
	int count_total = 0;
	int count_lines_removed = 0;

	while (fgets(buf, SHRT_MAX, trace_file) != nullptr) {
		line = buf;
		count_total++;
		if (line.find("fd") != std::string::npos) {
			count_lines_removed++;
			continue;
		}
		if (line.find("address") != std::string::npos) {
			count_lines_removed++;
			continue;
		}
		if (line.find("fildes") != std::string::npos) {
			count_lines_removed++;
			continue;
		}
		if (line.find("\"start\"") != std::string::npos) {
			count_lines_removed++;
			continue;
		}
		if (line.find("\"name\"") != std::string::npos) {
			count_lines_removed++;
			continue;
		}

		fputs(buf, clean_file);
	}

	fclose(trace_file);
	fclose(clean_file);
	fprintf(stderr, "Removed %d lines of %d total lines: %s\n",
	        count_lines_removed, count_total, clean_filename.c_str());

	return 0;
}

int tracer(int argc, char *argv[], bool retrace)
{
	char *exec[argc];
	int exec_index = 0;

	char retrace_command[] = "__retrace";

	if (retrace) {
		std::string trace_file = argv[optind];
		if (trace_file.find(".json") == std::string::npos) {
			line_info("\n\tTrace file \'%s\' must have .json file extension",
			          trace_file.c_str());
			print_usage();
			return -1;
		}
	}

	/* Get the application to be traced. */
	if (retrace) {
		exec[exec_index++] = argv[0]; /* tracee is v4l2-tracer, local or installed */
		exec[exec_index++] = retrace_command;
		exec[exec_index++] = argv[optind]; /* json file to be retraced */
	} else {
		while (optind < argc)
			exec[exec_index++] = argv[optind++];
	}
	exec[exec_index] = nullptr;

	/* Create a unique trace filename and open a file. */
	std::string trace_id;
	if (retrace) {
		std::string json_file_name = argv[optind];
		trace_id = json_file_name.substr(0, json_file_name.find(".json"));
		trace_id += "_retrace";
	} else {
		const int timestamp_start_pos = 1;
		trace_id = std::to_string(100000 + time(nullptr) % 100000);
		trace_id = trace_id.substr(timestamp_start_pos) + "_trace";
	}
	setenv("TRACE_ID", trace_id.c_str(), 0);
	std::string trace_filename = trace_id + ".json";
	FILE *trace_file = fopen(trace_filename.c_str(), "w");
	if (trace_file == nullptr) {
		fprintf(stderr, "Could not open trace file: %s\n", trace_filename.c_str());
		perror("");
		return errno;
	}

	/* Open the json array.*/
	fputs("[\n", trace_file);

	/* Add v4l-utils package and git info to the top of the trace file. */
	std::string json_str;
	json_object *v4l2_tracer_info_obj = json_object_new_object();
	json_object_object_add(v4l2_tracer_info_obj, "package_version",
	                       json_object_new_string(PACKAGE_VERSION));
	std::string git_commit_cnt = STRING(GIT_COMMIT_CNT);
	git_commit_cnt = git_commit_cnt.erase(0, 1); /* remove the hyphen in front of git_commit_cnt */
	json_object_object_add(v4l2_tracer_info_obj, "git_commit_cnt",
	                       json_object_new_string(git_commit_cnt.c_str()));
	json_object_object_add(v4l2_tracer_info_obj, "git_sha",
	                       json_object_new_string(STRING(GIT_SHA)));
	json_object_object_add(v4l2_tracer_info_obj, "git_commit_date",
	                       json_object_new_string(STRING(GIT_COMMIT_DATE)));
	json_str = json_object_to_json_string(v4l2_tracer_info_obj);
	fwrite(json_str.c_str(), sizeof(char), json_str.length(), trace_file);
	fputs(",\n", trace_file);
	json_object_put(v4l2_tracer_info_obj);

	/* Add v4l2-tracer command line to the top of the trace file. */
	json_object *tracee_obj = json_object_new_object();
	std::string tracee;
	for (int i = 0; i < argc; i++) {
		tracee += argv[i];
		tracee += " ";
	}
	json_object_object_add(tracee_obj, "Trace", json_object_new_string(tracee.c_str()));
	const time_t current_time = time(nullptr);
	json_object_object_add(tracee_obj, "Timestamp", json_object_new_string(ctime(&current_time)));

	json_str = json_object_to_json_string(tracee_obj);
	fwrite(json_str.c_str(), sizeof(char), json_str.length(), trace_file);
	fputs(",\n", trace_file);
	json_object_put(tracee_obj);
	fclose(trace_file);

	/*
	 * Preload the libv4l2tracer library. The libv4l2tracer is looked up next to
	 * the executable first in order to support uninstalled build.
	 */
	std::string libv4l2tracer_path;
	std::string program = argv[0];
	std::size_t idx = program.rfind("/");
	struct stat sb;

	if (idx == std::string::npos)
		idx = 0;
	else
		idx++;

	libv4l2tracer_path = program.replace(program.begin() + idx, program.end(), "libv4l2tracer.so");

	if (stat(libv4l2tracer_path.c_str(), &sb) == -1) {
		/* If not found, get the libv4l2tracer library from the meson install path 'prefix' */
		libv4l2tracer_path = std::string(LIBTRACER_PATH) + "/libv4l2tracer.so";

		/* Otherwise, guess where the library might be for a cross-build. */
		if (stat(libv4l2tracer_path.c_str(), &sb) == -1) {
			std::size_t idx =  libv4l2tracer_path.find("/home");
			libv4l2tracer_path = libv4l2tracer_path.substr(idx);

			/* Finally, check if the user set a custom path using LD_PRELOAD. */
			if (stat(libv4l2tracer_path.c_str(), &sb) == -1) {
				if (getenv("LD_PRELOAD"))
					libv4l2tracer_path = std::string(getenv("LD_PRELOAD"));

				if (stat(libv4l2tracer_path.c_str(), &sb) == -1) {
					fprintf(stderr, "Exiting: can't find libv4l2tracer library\n");
					fprintf(stderr, "Set a custom libv4l2tracer library path using: LD_PRELOAD\n");
					exit(EXIT_FAILURE);
				}
			}
		}
	}

	if (is_verbose())
		fprintf(stderr, "Loading libv4l2tracer: %s\n", libv4l2tracer_path.c_str());
	setenv("LD_PRELOAD", libv4l2tracer_path.c_str(), 0);

	tracee_pid = fork();
	if (tracee_pid == 0) {

		if (is_debug()) {
			line_info();
			fprintf(stderr, "\ttracee: ");
			for (int i = 0; i < exec_index; i++)
				fprintf(stderr,"%s ", exec[i]);
			fprintf(stderr, "\n");
		}

		execvpe(exec[0], (char* const*) exec, environ);
		line_info("\n\tCould not execute application \'%s\'", exec[0]);
		perror(" ");
		return errno;
	}

	int exec_result = 0;
	wait(&exec_result);

	if (WIFEXITED(exec_result))
		WEXITSTATUS(exec_result);

	fprintf(stderr, "Tracee exited with status: %d\n", exec_result);

	/* Close the json-array and the trace file. */
	trace_file = fopen(trace_filename.c_str(), "r+");
	fseek(trace_file, -2L, SEEK_END);
	fputs("\n]\n", trace_file);
	fclose(trace_file);

	if (retrace)
		fprintf(stderr, "Retrace complete: ");
	else
		fprintf(stderr, "Trace complete: ");
	fprintf(stderr, "%s", trace_filename.c_str());
	fprintf(stderr, "\n");

	unsetenv("LD_PRELOAD");
	return exec_result;
}

int main(int argc, char *argv[])
{
	int ret = -1;

	if (argc <= 1) {
		print_usage();
		return ret;
	}

	ret = get_options(argc, argv);

	if (ret < 0) {
		debug_line_info();
		return ret;
	}

	if (optind == argc) {
		debug_line_info();
		print_usage();
		return ret;
	}

	std::string command = argv[optind++];

	if (optind == argc) {
		debug_line_info();
		print_usage();
		return ret;
	}

	struct sigaction act = {};
	act.sa_handler = v4l2_tracer_sig_handler;
	sigaction(SIGINT, &act, nullptr);
	sigaction(SIGTERM, &act, nullptr);

	if (command == "trace") {
		ret = tracer(argc, argv);
	} else if (command == "retrace") {
		ret = tracer(argc, argv, true);
	} else if (command == "__retrace") {
		/*
		 * This command is meant to be used only internally to allow
		 * v4l2-tracer to recursively trace itself during a retrace.
		 */
		ret = retrace(argv[optind]);
	} else if (command == "clean") {
		ret = clean (argv[optind]);
	} else {
		if (is_debug()) {
			line_info("Invalid command");
			for (int i = 0; i < argc; i++)
				fprintf(stderr,"%s ", argv[i]);
			fprintf(stderr, "\n");
		}
		print_usage();
	}

	return ret;
}
